### ISSUE TITLE
Overload mismatch when type of an argument has a contravariant type-var

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pylance-release#2095]
## Environment data

-   Language Server version: v2021.11.3-pre.1
-   OS and version: Windows 7 x64
-   Python version (& distribution if applicable, e.g. Anaconda): 3.8.12 Anaconda

## Sample

```python
from __future__ import annotations

from typing import Generic, TypeVar, overload

_T = TypeVar("_T")
_T_contra = TypeVar("_T_contra", contravariant=True)

class C(Generic[_T_contra]):
    ...

@overload
def f(x: C[_T], /) -> _T:
    ...

@overload
def f(x: C[_T], /) -> _T:
    ...

def f(x: C[_T], /) -> _T:
    ...
```

## Expected behavior

No errors / warnings.

## Actual behaviour

![image](https://user-images.githubusercontent.com/15389744/142813478-abf4c233-a2f2-4bd2-95ba-e83b1998823f.png)

**Comment by Azureblade3808** [comment]
A more practial example would be
```python
from __future__ import annotations

from collections.abc import Iterable
from typing import Callable, TypeVar, overload

_T0 = TypeVar("_T0")
_T1 = TypeVar("_T1")

@overload
def my_map(func: Callable[[_T0], _T1], iterable: Iterable[_T0], /) -> _T1:
    ...

@overload
def my_map(func: Callable[[_T0], _T1], iterable: Iterable[_T0], /) -> _T1:
    ...

def my_map(func: Callable[[_T0], _T1], iterable: Iterable[_T0], /) -> _T1:
    ...
```

**Comment by erictraut** [comment]
I think this caused by the same as the bug described in this issue: https://github.com/microsoft/pyright/issues/2568.

**Comment by Azureblade3808** [comment]
I'd suggest that Pylance/Pyright treat the original sample as invalid, because there is no way to properly implement the function `f`.

And as a result of current behavior, with additional code like
```python
class Ancestor: ...
class Parent(Ancestor): ...
class Child(Parent): ...

x: C[Parent] = globals()[""]
y = f(x)  # Inferred as `Parent`.
y0: Parent = f(x)  # Valid.
y1: Ancestor = f(x)  # Valid.
y2: Child = f(x)  # Still valid?
```
, assignments for `y0`, `y1` and `y2` are all considered valid, which is a supprise.

I think the key point is, in a function signature, a contravariant type-var must be properly restricted into an invariant type-var, just as the second sample shows.

**Comment by Azureblade3808** [comment]
> I think this caused by the same as the bug described in this issue: [microsoft/pyright#2568](https://github.com/microsoft/pyright/issues/2568).

As a side note, I don't think it is wrong to emit a warning for the code there, because `Iterable[T] | T` is indeed ambiguous. So is the case in my original sample.

**Comment by Azureblade3808** [comment]
@eriictraut I found that microsoft/pyright#2568 has been closed, and see your reply there. But I think the problem I've faced is a bit different.

Here is my real code:
```python
from __future__ import annotations

from collections.abc import Iterable
from typing import Any, Callable, TypeVar, overload

_UNSPECIFIED: Any = object()

_T0 = TypeVar("_T0")
_T1 = TypeVar("_T1")
_T2 = TypeVar("_T2")

@overload
def try_map(func: Callable[[_T0], _T1], iterable: Iterable[_T0], /) -> Iterable[_T1]:
    ...

@overload
def try_map(
    func: Callable[[_T0], _T1], iterable: Iterable[_T0], /, default_value: _T2
) -> Iterable[_T1 | _T2]:
    ...

def try_map(
    func: Callable[[_T0], _T1],
    iterable: Iterable[_T0],
    /,
    default_value: _T2 = _UNSPECIFIED,
) -> Iterable[_T1 | _T2]:
    """
    A function which is similar to the built-in `map` function, except
    for that the first argument may fail.

    When the mapping function fails on an input value, a default value
    will take place of the output value if it is provided, otherwise this
    input value is just ignored.
    """

    # Implementation here.
    ...
```

As you can see from the error messages, the implementation is considered not consistent with either overload.

To make things clear, I tried to replace `Callable` with a custom version, and the code would become:
```python
from __future__ import annotations

from collections.abc import Iterable
from typing import Any, Generic, TypeVar, overload

_Input = TypeVar("_Input", contravariant=True)
_Output = TypeVar("_Output", covariant=True)

class Callable(Generic[_Input, _Output]):
    ...

_UNSPECIFIED: Any = object()

_T0 = TypeVar("_T0")
_T1 = TypeVar("_T1")
_T2 = TypeVar("_T2")

@overload
def try_map(func: Callable[_T0, _T1], iterable: Iterable[_T0], /) -> Iterable[_T1]:
    ...

@overload
def try_map(
    func: Callable[_T0, _T1], iterable: Iterable[_T0], /, default_value: _T2
) -> Iterable[_T1 | _T2]:
    ...

def try_map(
    func: Callable[_T0, _T1],
    iterable: Iterable[_T0],
    /,
    default_value: _T2 = _UNSPECIFIED,
) -> Iterable[_T1 | _T2]:
    """
    A function which is similar to the built-in `map` function, except
    for that the first argument may fail.

    When the mapping function fails on an input value, a default value
    will take place of the output value if it is provided, otherwise this
    input value is just ignored.
    """

    # Implementation here.
    ...
```

And error messages remain almost the same.

Then  I removed the `contravariant` part from `_Input`, and the errors were gone.

Sothe problem is with the contra-variance, right? But a type variable used in a function signature is always invariant itself, therefore I don't think a signature mismatch should exist here.

And what's more, `_T0` serves as both the input type of `Callable`, which should be contra-variant, and the element type of `Iterable`, which should be co-variant, so it should really be invariant.

**Comment by erictraut** [comment]
This is a separate issue. I agree this one is a bug. It will be addressed in the next release.

**Comment by bschnurr** [comment]
This issue has been fixed in version 2022.1.0, which we've just released. You can find the changelog here:  [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202210-5-january-2022)
